modules/services/borgmatic.nix

```
@@ -0,0 +1,298 @@
+              name = "repository";
+              frequency = "2 weeks";
+            }
+            {
+              name = "archives";
+              frequency = "4 weeks";
+            }
+            {
+              name = "data";
+              frequency = "6 weeks";
+            }
+            {
+              name = "extract";
+              frequency = "6 weeks";
+            }
+          ];
+        '';
+      };
+      extraConfig = extraConfigOption;
+    };
+  };
+
+  configModule = types.submodule {
+    options = {
+      location = mkOption {
+        type = locationModule;
+        default = { };
+        example = literalExpression ''
+          {
+            source_directories = [config.home.homeDirectory];
+            repositories = ["ssh://myuser@myserver.com/./personal-repo"];
+            extraConfig = {
+              one_file_system = true;
+              exclude_from = [./borg-excludes.txt];
+            };
+          }
+        '';
+      };
+      storage = mkOption {
+        type = storageModule;
+        default = { };
+      };
+      retention = mkOption {
+        type = retentionModule;
+        default = { };
+        example = literalExpression ''
+          {
+            keep_within = "2d";
+            keep_hourly = 2;
+            keep_daily = 7;
+            keep_weekly = 4;
+            keep_monthly = -1;
+            keep_yearly = -1;
+          }
+        '';
+      };
+      consistency = mkOption {
+        type = consistencyModule;
+        default = { };
+      };
+    };
+  };
+
+  # Move attrSet.extraConfig attributes to the top-level of attrSet:
+  inlineExtraConfig = attrSet:
+    (removeAttrs attrSet [ "extraConfig" ]) // attrSet.extraConfig;
+
+  removeNullValues = attrSet: filterAttrs (key: value: value != null) attrSet;
+
+  cleanupConfig = config: removeNullValues (inlineExtraConfig config);
+
+  writeConfig = config:
+    generators.toYAML { } {
+      location = cleanupConfig config.location;
+      storage = cleanupConfig config.storage;
+      retention = cleanupConfig config.retention;
+      consistency = cleanupConfig config.consistency;
+    };
+in {
+  meta.maintainers = [ maintainers.DamienCassou ];
+
+  options = {
+    services.borgmatic = {
+      enable = mkEnableOption "Borgmatic service";
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.borgmatic;
+        defaultText = "pkgs.borgmatic";
+        description = ''
+          The borgmatic package to run.
+        '';
+      };
+
+      frequency = mkOption {
+        type = types.str;
+        default = "hourly";
+        description = ''
+          How often to run borgmatic.  This value is passed to the systemd
+          timer configuration as the onCalendar option.  See
+          <citerefentry>
+            <refentrytitle>systemd.time</refentrytitle>
+            <manvolnum>7</manvolnum>
+          </citerefentry>
+          for more information about the format.
+        '';
```

Since this is the only option (beside the package) one that is direcly used by the timer service I would rename the module to `programs.borgmatic` and have an accompanying service module with a single `enable` option. Then can change this description to something like:

```suggestion
        description = ''
          How often to run borgmatic when
          <code language="nix">services.borgmatic.enable = true</code>.
          This value is passed to the systemd timer configuration as
          the onCalendar option. See
          <citerefentry>
            <refentrytitle>systemd.time</refentrytitle>
            <manvolnum>7</manvolnum>
          </citerefentry>
          for more information about the format.
        '';
```